File: leetcode/2001-3000/2901-3000/2961-2970/2962-count-subarrays-where-max-element-appears-at-least-k-times/CountSubarraysWhereMaxElementAppearsAtLeastKTimes.ts
function countSubarrays(numbers: number[], k: number): number {
  if (numbers.length === 0) {
    return 0;
  }

  if (k < 1) {
    return -1;
  }

  const maxElement = Math.max(...numbers);
  const totalElements = numbers.length;

  let totalMaxElements = 0;
  let totalSubarrays = 0;

  let i = 0;
  let j = 0;

  while (i < totalElements) {
    while (j < totalElements && totalMaxElements < k) {
      if (numbers[j] === maxElement) {
        totalMaxElements++;
      }

      // Prepare for the next round
      j++;
    }

    if (totalMaxElements < k) {
      break;
    }

    totalSubarrays += totalElements - j + 1;

    // Prepare for the next round
    if (numbers[i] === maxElement) {
      totalMaxElements--;
    }

    i++;
  }

  return totalSubarrays;
}

// ---------------------------------------------------------------------------
// Main function
// ---------------------------------------------------------------------------
const numbers = [1, 3, 2, 3, 3];
const k = 2;

console.log(countSubarrays(numbers, k));

File: leetcode/2001-3000/2901-3000/2961-2970/2962-count-subarrays-where-max-element-appears-at-least-k-times/CountSubarraysWhereMaxElementAppearsAtLeastKTimes.java
public class CountSubarraysWhereMaxElementAppearsAtLeastKTimes {
  public static void main(String[] args) {
    var numbers = new int[] { 1, 3, 2, 3, 3 };
    var k = 2;

    System.out.println((new Solution()).countSubarrays(numbers, k));
  }
}

class Solution {
  public long countSubarrays(int[] numbers, int k) {
    if (numbers.length == 0 || k < 1) {
      return 0;
    }

    int maxElement = numbers[0];
    for (int i = 1; i < numbers.length; i++) {
      if (numbers[i] > maxElement) {
        maxElement = numbers[i];
      }
    }

    long totalSubarrays = 0;
    int totalMaxElements = 0;

    int j = 0;

    for (int i = 0; i < numbers.length; i++) {
      while (totalMaxElements < k && j < numbers.length) {
        if (numbers[j] == maxElement) {
          totalMaxElements++;
        }

        j++;
      }

      if (totalMaxElements < k) {
        break;
      }

      totalSubarrays += numbers.length - j + 1;

      if (numbers[i] == maxElement) {
        totalMaxElements--;
      }

    }

    return totalSubarrays;
  }
}
